presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java

```diff
@@ -392,6 +409,40 @@ public void enforceQueryMaxRunTimeLimits()
     }
 
     /**
+     * Enforce max driver number at the query level
+     */
+    public void enforceQueryMaxTasksLimits()
+    {
+        for (QueryExecution query : queries.values()) {
+            if (query.getQueryInfo().getState().isDone()) {
+                continue;
+            }
+            int queryMaxDrivers = SystemSessionProperties.getQueryMaxDrivers(query.getSession());
+            int totalDrivers = query.getQueryInfo().getQueryStats().getTotalDrivers();
+            if (queryMaxDrivers < totalDrivers) {
+                query.fail(new PrestoException(EXCEEDED_DRIVER_NUMBER_LIMIT, "Query exceeded maximum driver number limit of " + queryMaxDrivers));
```

Comment:
The Hive connector doesn't know this before the query starts because it enumerates splits in a pipelined manner as the query executes. That's why `SELECT * FROM huge_table LIMIT 10` comes back immediately in Presto (and could take over an hour just to start executing in Hive).